simpleviewinc/keg-hub
# Keg publish repos

**Ticket**: [issue_id](https://jira.simpleviewtools.com/browse/ZEN-416)

## Notes
* further updates && cleanup items will be handled on a different ticket
    * https://jira.simpleviewtools.com/browse/ZEN-504 - git lib && output cleanup
    * https://jira.simpleviewtools.com/browse/ZEN-503 - tap publish


## Update (12/17/20)
* Added validation for the yarn script cmd
* updated test instruction for rollback. it will now switch back to the original branch and removes the generated release branch locally
* Went through all the feedback comments and updated / replied to them

## Context

* Whenever we make a new release for keg-hub. it usually requires very tedious steps that could easily be missed from user error
* we should automate this process to make life easier

## Goal

* Be able to publish our main keg repo packages with ease

## Updates

* `repos/keg-cli/package.json`
    * updated to use latest `jsutils` to use `noPropObj`
* `repos/keg-cli/scripts/setup/cli.config.json`
    * added default config for publish task
* `repos/keg-cli/src/libs/fileSys/fileSys.js`
    * add wrapper helper for recursively copy over folder contents
    * add wrapper helper for removing folder content (keeping the folder itself)
* `repos/keg-cli/src/tasks/hub/run.js`
    * utilized the yarnScript helper
* `repos/keg-cli/src/utils/helpers/runYarnScript.js`
   * helper to run script command at a given location
* `repos/keg-cli/src/tasks/hub/publish.js`
    * publish task with some options
* `repos/keg-cli/src/utils/hub/getHubRepos.js`
    * update to allow running scripts sync || async
* `repos/keg-cli/src/utils/publish/getPublishContext.js`
    * helper to get and merge the publish configs from the global config
* `repos/keg-cli/src/utils/publish/getPublishContextOrder.js`
   * ensure the order of the repos match the order defined in the config
* `repos/keg-cli/src/utils/services/publishService.js`
    * handles most of the logic for publish command
    * validates proper versioning
    * runs the `yarn` commands needed prior to publishing
    * creates a new git branch before commiting and pushing
* `repos/keg-cli/src/utils/services/versionService.js`
    * updates the version in package.json then updates the dependencies on other repos that depend on it
 

## Testing
* Pull this branch
* Run `keg cli && yarn install`
* Run `keg hub run test cli`
    * ensure all test passes
* Run `keg config sync`
    * *this will update your local keg config with the configs required to run publish cmds*


 **IMPORTANT**: we don't want to actually publish or push anything during testing, so make sure to always run `publish` command with `--dry-run`

### Publish single repo
* run `keg hub publish --context retheme --dry-run`
* for valid version, you can enter `minor/major/patch` or a valid version number: x.x.x
* Verify that no error appears and you see the `Publish Summary`
* run `keg hub && keg br` - verify you're now on a different branch `retheme-<new_version>`
* Open `keg-hub` in vscode and check the files changed
    * [image](https://user-images.githubusercontent.com/3317835/102248988-16f89e80-3ebf-11eb-971b-dcc330e8bb4d.png)
* Verify that retheme `package.json` version has been upgraded
* Verify that keg-components and keg-core `package.json` has been updated with the new version of retheme

### Publish dependent repos
* Clean git && switch back to this branch
* Make a change in `retheme` and `keg-components`
     * any change that can be seen when you run `keg-core` (i.e. console.log or component update)
     * for keg-components I updated `keg-components/src/theme/defaults.json` - [image](https://user-images.githubusercontent.com/3317835/102250959-6c35af80-3ec1-11eb-88fb-e8e698006dcc.png)
     * for re-theme I updated `re-theme/src/theme/index.js` - [image](https://user-images.githubusercontent.com/3317835/102251102-98e9c700-3ec1-11eb-859b-920b7569ce16.png)


* run `keg hub publish --context keg --dry-run`
     * this will:
           * upgrade `retheme > keg-components > keg-core` in that order
           * copy over dependee `/build` files the depender's `node_modules` before running `yarn build`
* Verify retheme, keg-components, keg-core `package.json` has been updated (app version and dependencies version)
* run `keg core && yarn web` 
* verify that you see the changes you've made from earlier steps
    * [image](https://user-images.githubusercontent.com/3317835/102251842-920f8400-3ec2-11eb-872d-1499a73b373d.png)
    * *this validates the `/build` copy over*

### Rollback on error
* Clean git && switch back to this branch
* open `keg-hub/repos/keg-components/src/utils/styles/__tests__/platformFlatten.js`
* modify one of the test case so that it would fail
   * [image](https://user-images.githubusercontent.com/3317835/102252476-5c1ecf80-3ec3-11eb-9b1c-c4d94db2be99.png)
* run `keg hub publish --context components --dry-run`
* It should prompt you for a `rollback` once the unit test fails
* accept the rollback
* verify that it switches back to your original branch and it deletes the release branch locally
     * [image](https://user-images.githubusercontent.com/3317835/102421246-3b3ca400-3fc1-11eb-82a4-b4e1792c892d.png) 
* verify that your working tree is clean - [image](https://user-images.githubusercontent.com/3317835/102252797-c5064780-3ec3-11eb-8e45-ec18c7ee83a3.png)
